fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneIndexMaintenanceTest.java

```diff
@@ -403,6 +434,314 @@ void flakyMerge(boolean isGrouped,
+        merger.start();
+
+        inserter.join();
+        merger.join();
+
+        // validate index is sane
+        new LuceneIndexTestValidator(() -> openContext(contextProps), context -> {
+            Pair<FDBRecordStore, QueryPlanner> storeAndPlanner = schemaSetup.apply(context);
+            return storeAndPlanner.getLeft();
+        }).validate(index, insertedDocs, Integer.MAX_VALUE, "text:about", false);
+    }
+
+    // A test where there are multiple threads trying to do merges. At the end the index should be validated for consistency.
+    @Test
+    void multipleConcurrentMergesTest() throws IOException {
+        final Map<String, String> options = Map.of(
+                INDEX_PARTITION_BY_FIELD_NAME, "timestamp",
+                INDEX_PARTITION_HIGH_WATERMARK, String.valueOf(100));
+
+        Index index = complexPartitionedIndex(options);
+
+        Function<FDBRecordContext, Pair<FDBRecordStore, QueryPlanner>> schemaSetup = context ->
+                LuceneIndexTestUtils.rebuildIndexMetaData(context, path, TestRecordsTextProto.ComplexDocument.getDescriptor().getName(), index, useCascadesPlanner);
+        final RecordLayerPropertyStorage contextProps = RecordLayerPropertyStorage.newBuilder()
+                .addProp(LuceneRecordContextProperties.LUCENE_REPARTITION_DOCUMENT_COUNT, 8)
+                .build();
+
+        final int countReps = 20;
+        final int threadCount = 10;
+
+        long timestamp = System.currentTimeMillis();
+        Map<Tuple, Map<Tuple, Tuple>> insertedDocs = new HashMap<>();
+
+        // create a bunch of docs
+        for (int i = 0; i < countReps; i++) {
+            try (FDBRecordContext context = openContext(contextProps)) {
+                Pair<FDBRecordStore, QueryPlanner> recordStoreQueryPlannerPair = schemaSetup.apply(context);
+                FDBRecordStore recordStore = recordStoreQueryPlannerPair.getLeft();
+                recordStore.getIndexDeferredMaintenanceControl().setAutoMergeDuringCommit(false);
+                TestRecordsTextProto.ComplexDocument cd = TestRecordsTextProto.ComplexDocument.newBuilder()
+                        .setGroup(1)
+                        .setDocId(i + 1000L)
+                        .setIsSeen(true)
+                        .setText("A word about what I want to say")
+                        .setTimestamp(timestamp + i)
+                        .setHeader(TestRecordsTextProto.ComplexDocument.Header.newBuilder().setHeaderId(1000L - i))
+                        .build();
+                Tuple primaryKey = recordStore.saveRecord(cd).getPrimaryKey();
+                insertedDocs.computeIfAbsent(Tuple.from(1), k -> new HashMap<>()).put(primaryKey, Tuple.from(timestamp));
+                commit(context);
+            }
+        }
+
+        List<Thread> mergingThreads = new ArrayList<>();
+        for (int i = 0; i < threadCount; i++) {
+            mergingThreads.add(new Thread(() -> explicitMergeIndex(index, contextProps, schemaSetup)));
```

Comment:
Since starting a thread takes time, it would probably be good to use a `CountDown` to increase the chance that the merges start at about the same time.